e2e/e2e_test.go

```diff
@@ -81,23 +97,253 @@ var _ = Describe("Whereabouts functionality", func() {
 			})
 
 			It("allocates a single pod within the correct IP range", func() {
+				const ipv4TestRange = "10.10.0.0/16"
+
 				By("checking pod IP is within whereabouts IPAM range")
 				secondaryIfaceIP, err := secondaryIfaceIPValue(pod)
 				Expect(err).NotTo(HaveOccurred())
 				Expect(inRange(ipv4TestRange, secondaryIfaceIP)).To(Succeed())
 			})
 		})
+
+		Context("Replicaset tests", func() {
+
+			var (
+				replicaCount int32
+				k8sClient    *kubeClient.Client
+				label        string
+				k8sIPAM      *kubeClient.KubernetesIPAM // - may be used for GetIPPools - todo
+			)
+
+			BeforeEach(func() {
+				By("creating a replicaset with whereabouts net-attach-def")
+
+				replicaCount = int32(0) // begin test with 0 replicas.
+				label = wbLabelEqual
+
+				k8sClient, err = kubeClient.NewClientViaKubeconfig(envVars.kubeconfigPath)
+				Expect(err).NotTo(HaveOccurred())
+
+				k8sIPAM, err = kubeClient.NewKubernetesIPAMWithNamespace("", types.IPAMConfig{
+					Kubernetes: types.KubernetesConfig{
+						KubeConfigPath: envVars.kubeconfigPath,
+					},
+				}, testNamespace)
+				Expect(err).NotTo(HaveOccurred())
+
+				replicaSet, err = clientInfo.provisionReplicaSet(
+					replicaCount,
+					podTierLabel(rsName),
+					podNetworkSelectionElements(testNetworkName),
+				)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				By("deleting replicaset with whereabouts net-attach-def")
+				Expect(clientInfo.deleteReplicaSet(replicaSet)).To(Succeed())
+			})
+
+			It("allocates each IP pool entry with a unique pod IP", func() {
+				const (
+					maxPodsPerNode  = 110
+					rsSteadyTimeout = 2 * rsCreateTimeout
+				)
+
+				By("creating max number of pods and checking IP Pool validity")
+				for i := 0; i < envVars.numThrashIter; i++ {
+					// Set to 0 replicas
+					replicaCount = int32(0)
+					replicaSet, err = clientInfo.updateReplicaSet(replicaSetObject(
+						replicaCount,
+						podTierLabel(rsName),
+						podNetworkSelectionElements(testNetworkName),
+					))
+					Expect(err).NotTo(HaveOccurred())
+
+					Expect(WaitForReplicaSetSteadyState(clientInfo.Client, testNamespace, label, replicaSet, rsSteadyTimeout)).To(Succeed())
+
+					// Check for 0 IP pool allocations
+					ipPool, err := k8sIPAM.GetIPPool(context.Background(), ipPoolName)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ipPool.Allocations()).To(BeEmpty())
+
+					// Increase to desired number of replicas for test
+					allPods, err := clientInfo.Client.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					replicaCount = int32((envVars.numComputeNodes*maxPodsPerNode - (len(allPods.Items))) * envVars.fillPercentCapacity / 100)
```

Comment:
instead of using this, I would create a function that returns this, and inline the usage. The trick here is the proper naming, for instance:

```golang
replicaSet, err = clientInfo.updateReplicaSet(replicaSetObject(
                                 envVars.maxReplicas(allPods.Items),
                                 podTierLabel(rsName),
                                 podNetworkSelectionElements(testNetworkName),
))

func (v envVars) maxReplicas(allPods []v1.Pod) int32 {
    const maxPodsPerNode  = 110
    return int32(
        (v.numComputeNodes*maxPodsPerNode - (len(allPods))) * v.fillPercentCapacity / 100)
}
```